libyara/scanner.c

```diff
@@ -466,59 +479,79 @@ YR_API int yr_scanner_scan_mem_blocks(
 
     block = iterator->next(iterator);
   }
+  // Come here if finished last block or no block currently available.
 
-  YR_TRYCATCH(
-      !(scanner->flags & SCAN_FLAGS_NO_TRYCATCH),
-      { result = yr_execute_code(scanner); },
-      { result = ERROR_COULD_NOT_MAP_FILE; });
+  YR_MEMORY_BLOCK* block_not_ready = iterator->first(iterator);
```

Comment:
This looks strange. Why calling `iterator->first()` here? Are you using `iterator->first()` for determining whether the previous call to  `iterator->next()` returned NULL because its the final block or because it was not ready?

If that's the case I think we need a better mechanism for knowing if `iterator->first()` and `iterator->next()` have reached the end of the block list or are indicating that the block is not available yet. Changing the signature of those functions so that they return a status code instead of a pointer to the block, or alternatively, adding a new field to the block that indicates if the block is actually ready or not.

Reply:
Yes, I used `iterator->first()` to determine if the last call to `iterator->next()` returning `NULL` means 'last block', or 'block not ready'. I agree with you that this can be made better.

As a more intention revealing mechanism, I submit part 2 which extends `YR_MEMORY_BLOCK_ITERATOR` with `last_error` which is set to `ERROR_SUCCESS` if `iterator->next()` returning `NULL` means 'last block', or is set to `ERROR_BLOCK_NOT_READY` if `iterator->next()` returning `NULL` means 'block not ready'.

Reply:
The solution in part 2 is much better.